crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1406,14 +1407,56 @@ impl<'db> TypeInferenceBuilder<'db> {
         let ast::StmtAugAssign {
             range: _,
             target,
-            op: _,
+            op,
             value,
         } = assignment;
-        self.infer_expression(value);
-        self.infer_expression(target);
 
-        // TODO(dhruvmanila): Resolve the target type using the value type and the operator
-        Type::Todo
+        // Resolve the target type, assuming a load context.
+        let target_type = match &**target {
+            Expr::Name(target) => self.infer_name_load(target),
+            Expr::Attribute(target) => self.infer_attribute_load(target),
+            _ => self.infer_expression(target),
+        };
+        let value_type = self.infer_expression(value);
+
+        // TODO(charlie): Add remaining branches for different types of augmented assignments.
+        if let (Operator::Sub, Type::Instance(class)) = (*op, target_type) {
```

Comment:
The test you added above is great, but it's not even testing this branch at all, since the type won't be `Type::Instance` but rather `Type::IntLiteral`, so you're getting the behavior from the fallback to regular binary ops below.

Let's add a test where we actually define `__isub__` on a type and show that we take its return type. And a test where we set `__isub__` to something not callable and show we get an error. (And the TODO test I mention below about `__isub__` not accepting `value_type` as argument.)

Reply:
Hah -- thank you, that's embarrassing :)